File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        if(head == nullptr) return head;
        Node* res = new Node(0);
        unordered_map<Node*, Node*> map;
        Node* cur = head;
        Node* curRes = res;
        int count = 0;
        while(cur){
            curRes->next = new Node(cur->val);
            curRes = curRes->next;
            map[cur] = curRes;
            cur = cur->next;
            count++;
        }
        cur = head;
        for(int i=0;i<count;i++){
            Node* temp = cur;
            map[temp]->random = map[temp->random];
            cur = cur->next;
        }
        return res->next;
    }
};

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if(head == null) return head;
        Map<Node, Node> mp = new HashMap<>();
        Node res = new Node(0);
        Node cur = res;
        Node temp = head;
        while(temp != null){
            cur.next = new Node(temp.val);
            cur = cur.next;
            mp.put(temp, cur);
            temp = temp.next;
        }
        temp = head;
        cur = res.next;
        while(temp != null){
            cur.random = mp.get(temp.random);
            temp = temp.next;
            cur = cur.next;
        }
        return res.next;
    }
}